### PR TITLE
integrate turbo workflow

### PR DESCRIPTION
**Pull request by PPpro**
Re: 
https://github.com/cocos/cocos-engine/issues/15911
https://github.com/cocos/cocos-engine/issues/16032

### Changelog

* move modules in scripts to pacakges directory
* integrate turbo workflow
* update tfig deps for build-engine

### before
once we need to fix a bug in build-engine, all we need to do is:
- fix bug in build-engine
- build and publish build-engine
- update deps in package.json
- run `npm i` to update deps

### now
for now, we only need to:
- fix bug in build-engine
- run `npm i`

`npm i` would install the workspaces dependencies automatically, and turbo would run the build task of each modules defined here https://github.com/cocos/cocos-engine/blob/7ec06307375a067a933bc6f71b367e33523e3786/packages/build-engine/package.json#L60
https://github.com/cocos/cocos-engine/blob/7ec06307375a067a933bc6f71b367e33523e3786/packages/native-pack-tool/package.json#L12
https://github.com/cocos/cocos-engine/blob/7ec06307375a067a933bc6f71b367e33523e3786/packages/typedoc-plugin/package.json#L8

### Relative PR

- https://github.com/cocos/creator-runtime-extensions/pull/479
-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+.turbo/
+node_modules/
 /bin
 /web.config
 .idea
 .history
 .DS_Store
 build
 aspnet_client
-node_modules
```
Comment (minggo) [inline .gitignore:9]: Why remove it?

Comment (PPpro) [reply .gitignore:9]: added here to recursively ignore node_modules in sub directories
https://github.com/cocos/cocos-engine/blob/55693df4c4116a76af0ba29b83d45d7f89677e31/.gitignore#L3

File: gulpfile.js [hunk 2]
```diff
@@ -39,7 +39,7 @@ gulp.task('build-debug-infos', async () => {
 });
 
 gulp.task('build-source', async () => {
-    const cli = require.resolve('@cocos/build-engine/dist/cli');
+    const cli = require.resolve('@cocos/build-engine/lib/cli');
```
Comment (minggo) [inline gulpfile.js:42]: Why change the url? Will it break compatibility?

Comment (PPpro) [reply gulpfile.js:42]: Turbo's build pipeline can implement build caching to improve build efficiency, provided that we need to configure the input and output of each module's build task

https://github.com/cocos/cocos-engine/blob/55693df4c4116a76af0ba29b83d45d7f89677e31/turbo.json#L4-L9

Here we unify the input of all modules into the src directory and the output into the lib directory

Comment (PPpro) [reply gulpfile.js:42]: it won't break compatibility, I also update the path deps for native-pack-tool here 
https://github.com/cocos/creator-runtime-extensions/pull/479

most of them are internal local modules, not public modules

**Review by minggo — APPROVED** [review]

**Review by PatriceJiang — APPROVED** [review]

**Review by shrinktofit — APPROVED** [review]

**Pull request merged**